### PR TITLE
Role management for groups and users

### PR DESCRIPTION
**Pull request by zxan1285**

### CODE REVIEW

File: client/components/Groups/GroupRoles.jsx [hunk 1]
```diff
@@ -1,23 +1,63 @@
...
-        </div>
-        <div className="col-xs-4">
-          <Button className="pull-right" bsStyle="success" onClick={() => {}}>
-            <i className="icon icon-budicon-473" /> Add role
-          </Button>
-        </div>
+        <Error message={error} />
+        <LoadingPanel show={loading}>
+          <GroupRolesDialog group={this.props.group}
+                            addRoles={this.props.addRoles}
```
Comment (sericaia) [inline client/components/Groups/GroupRoles.jsx]: can you please remove extra space here ?

File: client/reducers/groupRoles.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      record: null
+    }),
+  [constants.DELETE_GROUP_ROLE_PENDING]: (state) =>
+    state.merge({
+      ...initialState,
+      loading: true
+    }),
+  [constants.DELETE_GROUP_ROLE_REJECTED]: (state, action) =>
+    state.merge({
+      loading: false,
```
Comment (sericaia) [inline client/reducers/groupRoles.js]: shouldnt "deleting" be updated here?

File: client/components/UserGroupRoles/ItemRolesOverview.jsx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    applications: PropTypes.object.isRequired,
+    requestDeleteRole: PropTypes.func.isRequired,
+    cancelDeleteRole: PropTypes.func.isRequired,
+    deleteRole: PropTypes.func.isRequired
+  }
+
+  constructor() {
+    super();
+
+    this.searchBarOptions = [
```
Comment (sericaia) [inline client/components/UserGroupRoles/ItemRolesOverview.jsx]: This was copy paste, right?

File: client/components/UserGroupRoles/RolesTable.jsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import _ from 'lodash';
```
Comment (sericaia) [inline client/components/UserGroupRoles/RolesTable.jsx]: We dont need a duplicated RolesTable (there's an equal file in components/Roles folder)

File: client/actions/group.js [hunk 5]
```diff
@@ -217,3 +217,76 @@ export function closeDelete() {
...
+    }
+  };
+}
+
+export function fetchRolesForGroup(groupId) {
+  return {
+    type: constants.FETCH_GROUP_ROLES,
+    payload: {
+      promise: axios.get(`/api/groups/${groupId}/roles`, {
+        params: {},
```
Comment (sericaia) [inline client/actions/group.js]: this line can be removed if no params are sent ;)

File: client/actions/user.js [hunk 6]
```diff
@@ -84,3 +84,76 @@ export function fetchUser(userId) {
...
+    }
+  };
+}
+
+export function fetchRolesForUser(userId) {
+  return {
+    type: constants.FETCH_USER_ROLES,
+    payload: {
+      promise: axios.get(`/api/users/${userId}/roles`, {
+        params: {},
```
Comment (sericaia) [inline client/actions/user.js]: this line can be removed if no params are sent

File: client/components/Groups/GroupRoles.jsx [hunk 7]
```diff
@@ -1,23 +1,66 @@
...
+            itemId={this.props.groupId}
+          />
+        </LoadingPanel>
       </div>
     );
   }
 }
 
+GroupRoles.propTypes = {
+  roles: React.PropTypes.object,
```
Comment (sericaia) [inline client/components/Groups/GroupRoles.jsx]: add missing propTypes

File: client/components/Groups/GroupRolesDialog.jsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    const selectedIds = _.map(selected, role => role._id);
+    const options = _.filter(all, role => !_.includes(selectedIds, role._id)).map(role => ({
+      value: role._id,
+      text: role.name
+    }));
+
+    if (!options.length) {
+      return (
+        <Modal.Body>
+          <p className="modal-description">There are no roles you can add to this group.</p>
```
Comment (sericaia) [inline client/components/Groups/GroupRolesDialog.jsx]: I'm not sure if we should show this message or not. @sandrinodimattia may have an input on this.

File: client/reducers/userRoles.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      record: action.meta.role
+    }),
+  [constants.CANCEL_DELETE_USER_ROLE]: (state) =>
+    state.merge({
+      deleting: false,
+      record: null
+    }),
+  [constants.DELETE_USER_ROLE_PENDING]: (state) =>
+    state.merge({
+      ...initialState,
```
Comment (sandrinodimattia) [inline client/reducers/userRoles.js]: Set deleting to true here

File: client/containers/User.jsx [hunk 10]
```diff
@@ -72,15 +73,30 @@ export class UserContainer extends Component {
         </div>
         <div className="row">
           <div className="col-xs-12">
-            <Tabs defaultActiveKey={1} animation={false} style={{ marginTop: '20px' }}>
+            <Tabs defaultActiveKey={1} animation={false} style={{ marginTop: '20px' }} id="user_tabs">
```
Comment (sandrinodimattia) [inline client/containers/User.jsx]: user-tabs

File: client/containers/Group.jsx [hunk 11]
```diff
@@ -160,15 +160,30 @@ export class GroupContainer extends Component {
         </div>
         <div className="row">
           <div className="col-xs-12">
-            <Tabs defaultActiveKey={1} animation={false} style={{ marginTop: '20px' }}>
+            <Tabs defaultActiveKey={1} animation={false} style={{ marginTop: '20px' }} id="group_tabs">
```
Comment (sandrinodimattia) [inline client/containers/Group.jsx]: group-tabs

File: client/components/UserGroupRoles/ItemRolesOverview.jsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  render() {
+    const { error, loading, records, deleting, record } = this.props.roles.toJS();
+
+    if (loading) {
+      return this.renderLoading();
+    }
+
+    return (
+      <div>
+        {(record) ?
```
Comment (sandrinodimattia) [inline client/components/UserGroupRoles/ItemRolesOverview.jsx]: Don't do this inline logic. Use methods

File: client/components/Groups/GroupRolesDialog.jsx (hunk 8 above)
Comment (sandrinodimattia) [inline client/components/Groups/GroupRolesDialog.jsx]: Text here is fine

**Pull request merged**